x/stake/keeper/delegation.go

```diff
@@ -155,6 +155,57 @@ func (k Keeper) RemoveUnbondingDelegation(ctx sdk.Context, ubd types.UnbondingDe
 	store.Delete(GetUBDByValIndexKey(ubd.DelegatorAddr, ubd.ValidatorAddr))
 }
 
+// gets a specific unbonding queue timeslice. A timeslice is a slice of DVPairs corresponding to unbonding delegations
+// that expire at a certain time.
+func (k Keeper) GetUnbondingQueueTimeSlice(ctx sdk.Context, timestamp time.Time) (dvPairs []types.DVPair) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(GetUnbondingDelegationTimeKey(timestamp))
+	if bz == nil {
+		return []types.DVPair{}
+	}
+	k.cdc.MustUnmarshalBinary(bz, &dvPairs)
+	return dvPairs
+}
+
+// Sets a specific unbonding queue timeslice.
+func (k Keeper) SetUnbondingQueueTimeSlice(ctx sdk.Context, timestamp time.Time, keys []types.DVPair) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinary(keys)
+	store.Set(GetUnbondingDelegationTimeKey(timestamp), bz)
+}
+
+// Insert an unbonding delegation to the appropriate timeslice in the unbonding queue
+func (k Keeper) InsertUnbondingQueue(ctx sdk.Context, ubd types.UnbondingDelegation) {
+	timeSlice := k.GetUnbondingQueueTimeSlice(ctx, ubd.MinTime)
+	dvPair := types.DVPair{ubd.DelegatorAddr, ubd.ValidatorAddr}
+	if len(timeSlice) == 0 {
+		k.SetUnbondingQueueTimeSlice(ctx, ubd.MinTime, []types.DVPair{dvPair})
+	} else {
+		timeSlice = append(timeSlice, dvPair)
+		k.SetUnbondingQueueTimeSlice(ctx, ubd.MinTime, timeSlice)
+	}
+}
+
+// Returns all the unbonding queue timeslices from time 0 until endTime
+func (k Keeper) UnbondingQueueIterator(ctx sdk.Context, endTime time.Time) sdk.Iterator {
+	store := ctx.KVStore(k.storeKey)
+	return store.Iterator(UnbondingQueueKey, sdk.InclusiveEndBytes(GetUnbondingDelegationTimeKey(endTime)))
+}
+
+// Returns a concatenated list of all the timeslices before currTime, and deletes the timeslices from the queue
+func (k Keeper) GetAllMatureUnbondingQueue(ctx sdk.Context, currTime time.Time) (matureUnbonds []types.DVPair) {
```

Comment:
Doesn't seem like `GetAllMatureUnbondingQueue ` and `GetRedelegationQueueTimeSlice ` are idempotent. Based on the `Get*` nomenclature, I'd expect it to be so.

How do we feel about that? If we're going to keep it as non-idempotent. May I suggest changing the name?

Reply:
If I understand this correctly, during this `Get` we are actually also deleting the records from the queue after they are retrieved (aka modifying the state, aka non-idempotent) - Yeah, for that reason I'd agree we should consider a different function-name-prefix. Seems like the most technically correct term to use here would be `Dequeue` - ideally we should also prefix the function which adds elements to these queues with `Enqueue` to stay consistent

Reply:
++